rfcs/0000-open-arch.md

```diff
@@ -0,0 +1,225 @@
+  err = r.client.Update(ctx, rs)
+  if err != nil {
+    return reconcile.Result{}, fmt.Errorf("could not write MetaPolicy: %+v", err)
+  }
+
+  return reconcile.Result{}, nil
+}
+```
+It on differs in three ways from the original example:
+
+1. The `Reconcile` signature takes an additional parameter, the `kuadrant *KuadrantContext`;
+1. It acts upon an hypothetical `user.MetaPolicy`;
+1. It uses the `evaluateExpression` to resolve the CEL expression `kuadrant.gatewaysFor(self)[0].metadata.name` instead
+   of using a fixed string for the label's value.
+
+Because that expression is evaluated upon creation, the `KuadrantRuntime` will make sure to reconcile whenever that cel
+expression's evaluated result would change; in this particular example whenever the name of the first `Gateway` changes.
+
+> [!NOTE]
+> Below is the proposed inital integration point for `MetapolicyController`s. The idea is to keep the deployment model
+> fairly open moving forward. As of now, this would be the deployment model for _our own_ *Metapolicies*
+
+To deploy our custom reconciler, we need to let the Kuadrant operator know about it, by adding it to the `Kuadrant` CR:
+
+```yaml
+apiVersion: kuadrant.io/v1beta2
+kind: Kuadrant
+metadata:
+  name: kuadrant-sample
+spec:
+  plugins:
+    userMetaPolicy:
+      deployment: embedded
+```
+
+The `UserMetaPolicy` will need to be packaged with a custom Docker image containing the plugin. It'll be automatically
+registered with the Kuadrant Controller from its own config file.
+
+> [!TODO]
+> Support out-of-image deployments of plugins, i.e. ones running as their own container. Which then also means we need
+> to deal with failures et al, so probably adding the necessity for more configuration options for these plugins
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Plugin architecture
+
+### gRPC to controller
+
+- [ ] Unix socket for "in-pod/embedded" plugins
+- [ ] Declarative way to load plugins
+- [ ] Provide a decorated [Controller Runtime](https://github.com/kubernetes-sigs/controller-runtime) Golang that
+  provides the changes triggered by the DAG
+
+### gRPC streams for eventing on the DAG
+
+- [ ] Subscriptions infered from CEL and the current CR being reconcilied
+
+### Typed DSL in CEL
+
+> [!IMPORTANT]
+> Provide both the `MetapolicyCR` & the `GwAPIContext` instances relative to a reconciliation
+> The `GwAPIContext` would not only be responsible for resolving the [CEL expressions][3] but also deriving the
+> necessary subscriptions needed to inform the controller when a reconciliation is needed because of changes to the
+> Gateway API objects queried
+
+- [ ] Support existing `context`s in the CEL interpreter's `Env`
+- [ ] Find a way to express the `context`s present/set for each CEL
+
+
+### Deployment
+
+> [!IMPORTANT]
+> This would initially be deployed as their own processes running alongside the kuadrant controller. They'd communicate
```

Comment:
When you say along side. Does that mean in the same pod and as part of the same deployment object?

Reply:
One consideration may be limiting the resources that such a container can have. I think we would also want to understand to protect the kuadrant operator from a bad plugin crashing and potentially causing the pod to get restarted

Reply:
https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#container-restarts